src/utils/upload/sampleInspector.js

```diff
@@ -0,0 +1,136 @@
+import {
+  DecodeUTF8, Decompress,
+} from 'fflate';
+
+const Verdict = {
+  INVALID_BARCODES_FILE: 'INVALID_SAMPLE_FILES',
+  INVALID_FEATURES_FILE: 'INVALID_FEATURES_FILE',
+  INVALID_TRANSPOSED_MATRIX: 'INVALID_TRANSPOSED_MATRIX',
+};
+
+const verdictText = {
+  [Verdict.INVALID_BARCODES_FILE]: 'Barcodes file is invalid',
+  [Verdict.INVALID_FEATURES_FILE]: 'Features file is invalid',
+  [Verdict.INVALID_TRANSPOSED_MATRIX]: 'Matrix file is transposed',
+};
+
+const CHUNK_SIZE = 2 ** 18; // 250 kb
+
+const decode = async (arrBuffer) => {
+  let result = '';
+  const utfDecode = new DecodeUTF8((data) => { result += data; });
+  utfDecode.push(new Uint8Array(arrBuffer));
+
+  return result;
+};
+
+const decompress = async (arrBuffer) => {
+  let result = '';
+  const decompressor = new Decompress((chunk) => { result = chunk; });
+  decompressor.push(new Uint8Array(arrBuffer));
+
+  return result;
+};
+
+const extractSampleSizes = async (matrix) => {
+  const { compressed, fileObject } = matrix;
+  let header = '';
+  let matrixHeader = '';
+
+  const fileArrBuffer = await fileObject.slice(0, 500).arrayBuffer();
+
+  matrixHeader = compressed
+    ? await decode(await decompress(fileArrBuffer))
+    : await decode(fileArrBuffer);
+
+  // The matrix header is the first line in the file that splits into 3
+  header = matrixHeader.split('\n').find((line) => line.split(' ').length === 3);
+
+  const [featuresSize, barcodeSize] = header.split(' ');
```

Comment:
You are not checking that the number of lines in the matrix matches the header (the 3rd parameter in the header). I have tried to just remove lines from the bottom and it passes the validation.